src/Std/Sat/CNF/Basic.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kim Morrison
+-/
+prelude
+import Init.Data.List.Lemmas
+import Std.Sat.CNF.Literal
+
+namespace Std
+namespace Sat
+
+/--
+A clause in a CNF.
+
+The literal `(i, b)` is satisfied if the assignment to `i` agrees with `b`.
+-/
+abbrev CNF.Clause (α : Type u) : Type u := List (Literal α)
+
+/--
+A CNF formula.
+
+Literals are identified by members of `α`.
+-/
+abbrev CNF (α : Type u) : Type u := List (CNF.Clause α)
+
+namespace CNF
+
+/--
+Evaluating a `Clause` with respect to an assignment `a`.
+-/
+def Clause.eval (a : α → Bool) (c : Clause α) : Bool := c.any fun (i, n) => a i == n
+
+@[simp] theorem Clause.eval_nil (a : α → Bool) : Clause.eval a [] = false := rfl
+@[simp] theorem Clause.eval_cons (a : α → Bool) :
+    Clause.eval a (i :: c) = (a i.1 == i.2 || Clause.eval a c) := rfl
+
+/--
+Evaluating a `CNF` formula with respect to an assignment `a`.
+-/
+def eval (a : α → Bool) (f : CNF α) : Bool := f.all fun c => c.eval a
+
+@[simp] theorem eval_nil (a : α → Bool) : eval a [] = true := rfl
```

Comment:
Currently you develop half of the theory about clauses and CNFs in terms of `⊨`/`Unsatisfiable` and the other half in terms of the two `eval` methods. This is generally considered to be subobtimal design, because it makes the API difficult to discover and requires knowing when to magically unfold using `simp [(· ⊨ ·)]` to jump from one layer to the next (which then magically has completely different simp normal forms!). It would be much nicer if there was only one theorem `eval_def` for each of the `eval`s and all of the theory (like `eval_nil,` `eval_cons`, and also downstream things like `eval_relabel` are stated in terms of `⊨` instead of `eval`.